### DISCUSSION TITLE
How to prevent a tuple from being type-narrowed?

### DISCUSSION THREAD

**Discussion by xavdid** [microsoft/pyright#4317] [Q&A] [answered]
I've got a function decorator that verifies the return value of a function:

```py
def should_return(answer):
    def deco(func):
        @wraps(func)
        def wrapper(self):
            result = func(self)
            assert result == answer
            return result
        return wrapper
    return deco
```

It's used on instance methods of a specific class that don't take additional arguments (not for general use), so it doesn't need to worry about `*args`, etc. Here's the usage:

```py
class MyClass:
    @should_return(3)
    def something(self):
        return 3
```

It works great at runtime; the problem is the typing. The functions it wraps usually return `int` or `Tuple[int, int]`. But, occasionally there's a `str` in one or both parts of the tuple, so I'd like the types to be flexible.

Ideally, I could use a generic to sync up the return type of the wrapped function and the argument expected in `should_return`. Here's the code again, with typing: 

```py
from functools import wraps
from typing import Callable, Tuple, TypeVar

T = TypeVar("T")

def should_return(answer: T):
    def deco(func: Callable[["MyClass"], T]):
        @wraps(func)
        def wrapper(self):
            result = func(self)
            assert result == answer
            return result
        return wrapper
    return deco
```

This works great on the class, _unless_ there's a tuple return. I've read about this behavior [in the docs](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#tuple-expressions) and in many other discussions here, so I realize it's not a bug. But, it is a little frustrating here.

```py
class MyClass:
    # wrapped function says it returns an int,
    # so `should_return` expects an int
    # correct no error
    @should_return(3)
    def something(self) -> int:
        return 3

    # wrapped function is says it returns an int,
    # so `should_return` correctly has a type error (and eventually, a runtime error too)
    @should_return("3")
    def something_else(self) -> int:
        return 3

    # this should have no type error and no runtime error
    # but, the type is narrowed to Tuple[Literal[3], Literal[3]]
    # so there's a mismatch: 
    #   "int" cannot be assigned to type "Literal[3]"
    @should_return((3, 3))
    def something_tuple(self) -> Tuple[int, int]:
        return 3, 3
```

Pyright doesn't have a problem with `return 3, 3` lining up with the return type (even if I assign it to a variable that's inferred as a pair of literals first):

```py
def something_tuple(self) -> Tuple[int, int]:
    t = 3, 3 # inferred as `tuple[Literal[3], Literal[3]]`
    return t # no error
```

---

So, is there any way for pyright to accept a tuple shape in `should_return` that conforms to the return type of its wrapped function?

Thank you!

**Comment by erictraut** [comment, marked as answer]
The problem here is that pyright retains literal types when inferring the type of a tuple expression. So the inferred type of `(3, 3)` is `tuple[Literal[3], Literal[3]]`. Retaining literals for tuples is beneficial most of the time, but in this case it causes problems.

Here are a couple of potential solutions...

1. Since `something_tuple` always returns `(3, 3)`, its return type could be annotated as `tuple[Literal[3], Literal[3]]`, and that eliminates the type violation. I'm guessing that your actual use case isn't this simple, however. So this is probably an unsatisfactory answer.

2. You could use the new `TypeVarTuple` introduced in Python 3.11 to create a `make_tuple` function that does not retain literal types. You can use `typing_extensions` for backward compatibility with Python <3.11. Here's how that would look:

```python
from typing_extensions import TypeVarTuple, Unpack
Ts = TypeVarTuple("Ts")

def make_tuple(*args: Unpack[Ts]) -> tuple[Unpack[Ts]]:
    return tuple(*args)

...
    @should_return(make_tuple(3, 3))
    def something_tuple(self) -> tuple[int, int]:
        return 3, 3
```

**Reply by xavdid** [reply]
@erictraut thank you for the detailed answer!

1. `something_tuple` always returns the same value (as long as my code is working), but, you're correct that the containing code isn't so simple. It returns the result of a calculation, which is stored in a variable. So if I type the function as returning a tuple of literals, `@should_return` shows no type error, but the return complains the other way: `"int" cannot be assigned to type "Literal[3]"`
2. It does look like this solution will work, at the expense of adding the `make_tuple` call to any tuples passed into `@should_return`. Is there a way to type `@should_return` that would alleviate that call? I was playing with `def should_return(*answer: Ts):`, but couldn't quite get it working. I'm happy for `@should_return` to do some extra work to simplify its (many) callsites

Thank you for your time!

**Reply by erictraut** [reply]
You might be able to use `@overload` to make this work. Here's what that might look like:

```python
@overload
def should_return(answer: tuple[*Ts]) -> Callable[[Callable[[C], tuple[*Ts]]], Callable[[C], tuple[*Ts]]]: ...
@overload
def should_return(answer: T) -> Callable[[Callable[[C], T]], Callable[[C], T]]: ...
def should_return(answer: T) -> Callable[[Callable[[C], T]], Callable[[C], T]]:
    def deco(func: Callable[[C], T]) -> Callable[[C], T]:
        ...
```

**Reply by xavdid** [reply]
Ah, that's perfect! Thank you, you're a wizard.
